---
title: 记一次疑似内存泄漏故障
date: 2020-08-06 16:14:35
tags: [debug]
---
&emsp;&emsp;起因是一次寻常的开发调试过程，我像往常一样在开发完成后开始自测。

&emsp;&emsp;此时出现了一些奇怪的迹象。
<!--more-->
&emsp;&emsp;在分支逻辑中打印输出，在请求端却没有任何返回值。

&emsp;&emsp;起初，我以为是这天杀的tp框架配置的不对。因为我接手的是一个老旧项目，2017年的代码，用的是tp3.2.2框架。在开发过程中我就觉得这框架问题不少。可能是我的配置问题导致框架使用了生产环境模式。于是框架不输出报错页面，而是输出报错到日志了。

&emsp;&emsp;于是我翻文档，去搜索引擎查，看看是不是我的配置有问题。

&emsp;&emsp;调了半天，并没有结果。

&emsp;&emsp;这时我仔细想想。我初步得出的结论实际上是很有问题的。
1. 我查看了最近的修改记录，并没有修改相关的配置。在配置文件没有变化的情况下，不太可能会导致我目前遇到的问题
2. 此处的问题不是不输出报错，而是普通的打印都没有输出。
3. 即使是日志，我也没看到任何输出。

&emsp;&emsp;教训来了，debug的时候应该优先结合上下文来调试，以固有思维得出的结论很可能是不正确的。

&emsp;&emsp;继续想办法。

&emsp;&emsp;目前的情况是，不论日志还是打印都没有正确输出。

&emsp;&emsp;此时我猜测，也许是框架内部的错误处理逻辑有问题，导致了此次故障。

&emsp;&emsp;于是我找到了最初注册错误处理函数的代码。框架核心代码。

```
// 设定错误和异常处理
register_shutdown_function('Think\Think::fatalError');
set_error_handler('Think\Think::appError');
set_exception_handler('Think\Think::appException');
```

&emsp;&emsp;接下来我在这三个方法的开始都打印断点输出。发现在fatalError方法是有输出的，于是来到这个方法。

```
// 致命错误捕获
static public function fatalError() {echo json_encode(11111);
    Log::save();
    if ($e = error_get_last()) {
        switch($e['type']){
            case E_ERROR:
            case E_PARSE:
            case E_CORE_ERROR:
            case E_COMPILE_ERROR:
            case E_USER_ERROR:  
            ob_end_clean();
            self::halt($e);
            break;
        }
    }
}
```
&emsp;&emsp;其中这个 echo 是我打印的输出。

&emsp;&emsp;确定到这里就好办了，我打印错误具体内容。
```
// 致命错误捕获
static public function fatalError() {echo json_encode(error_get_last());
    Log::save();
    if ($e = error_get_last()) {
        switch($e['type']){
            case E_ERROR:
            case E_PARSE:
            case E_CORE_ERROR:
            case E_COMPILE_ERROR:
            case E_USER_ERROR:  
            ob_end_clean();
            self::halt($e);
            break;
        }
    }
}
```
&emsp;&emsp;内容如下
```
Allowed memory size of 134217728 bytes exhausted (tried to allocate 262144 bytes)
```

&emsp;&emsp;非常明显，说我耗尽了配置中允许的内存。

&emsp;&emsp;看到这里我非常疑惑，php在cgi模式下很难出现内存不够的问题，以往碰到这种情况，都是要导出超大文件，或者一次性读入太多数据的时候发生的。

&emsp;&emsp;可是我非常清楚本次出问题的代码没有这些场景。

&emsp;&emsp;为了验证猜测我特意将配置文件改为不限制内存，如果只是暂时性的内存不够，那么这样问题就会消失。

&emsp;&emsp;改完之后的报错如下
```
Out of memory (allocated 2057306112) (tried to allocate 262144 bytes)
```

&emsp;&emsp;呵呵，直接吃光了我能分配给他的所有内存，还说不够。

&emsp;&emsp;这时候可以确定目前的代码有内存泄漏的问题。

&emsp;&emsp;定位到最后，发现是一句加锁操作导致的问题。

```
$result = $this->where([
            "id" => $id,
        ])->lock(true)->find();
```

&emsp;&emsp;去掉这个锁操作就一切如常。

&emsp;&emsp;然后我朝源码跟了半天，并没有什么头绪，层层叠叠的魔术方法，不知道谁调用的谁。

&emsp;&emsp;不过最后我灵光一闪，找到了原因。

&emsp;&emsp;这段方法原文是这样的。
```
public function lock($id) {
    $result = $this->where([
        "id" => $id,
    ])->lock(true)->find();
    
    return $result;
}
```

&emsp;&emsp;因为这个方法名重写了父类里的lock方法。所以变成了循环调用，自然内存不够了。

### 总结
&emsp;&emsp;花了这么久，结果是自己的bug。

&emsp;&emsp;发现问题时一定结合现场上下文进行分析。过往的经验有时候会误导你的判断，以案发现场为准，层层递进，才是比较稳妥的方式。

&emsp;&emsp;如果选用了自己不信任的工具，造成的问题还是挺大的，出现问题时，不可信任的工具往往会让你倾向于把锅甩给它，而想要确认工具没有问题又是个成本高昂的选择，这就很难了。
